java/src/main/native/src/TableJni.cpp

```diff
@@ -22,12 +22,185 @@
+    const char * copy_from = static_cast<const char*>(data);
+    while (left_to_copy > 0) {
+      long buffer_amount_available = current_buffer_len - current_buffer_written;
+      if (buffer_amount_available <= 0) {
+        // should never be < 0, but just to be safe
+        rotate_buffer(env);
+        buffer_amount_available = current_buffer_len - current_buffer_written;
+      }
+      long amount_to_copy = left_to_copy < buffer_amount_available ? left_to_copy : buffer_amount_available;
+      char * copy_to = current_buffer_data + current_buffer_written;
+
+      std::memcpy(copy_to, copy_from, amount_to_copy);
+      copy_from = copy_from + amount_to_copy;
+      current_buffer_written += amount_to_copy;
+      total_written += amount_to_copy;
+      left_to_copy -= amount_to_copy;
+    }
+  }
+
+  bool supports_device_write() const override {
+    return true;
+  }
+
+  void device_write(void const* gpu_data, size_t size, cudaStream_t stream) {
+    if (this->stream != stream) {
+      if (is_stream_set) {
+        CUDA_TRY(cudaStreamSynchronize(this->stream));
+      }
+      this->stream = stream;
+    }
+    this->is_stream_set = true;
+
+    JNIEnv* env = cudf::jni::get_jni_env(jvm);
+    size_t left_to_copy = size;
+    const char * copy_from = static_cast<const char*>(gpu_data);
+    while (left_to_copy > 0) {
+      long buffer_amount_available = current_buffer_len - current_buffer_written;
+      if (buffer_amount_available <= 0) {
+        // should never be < 0, but just to be safe
+        rotate_buffer(env);
+        buffer_amount_available = current_buffer_len - current_buffer_written;
+      }
+      long amount_to_copy = left_to_copy < buffer_amount_available ? left_to_copy : buffer_amount_available;
+      char * copy_to = current_buffer_data + current_buffer_written;
+
+      CUDA_TRY(cudaMemcpyAsync(copy_to, copy_from, amount_to_copy, cudaMemcpyDeviceToHost, stream));
+
+      copy_from = copy_from + amount_to_copy;
+      current_buffer_written += amount_to_copy;
+      total_written += amount_to_copy;
+      left_to_copy -= amount_to_copy;
+    }
+  }
+
+  void flush() override {
+    if (current_buffer_written > 0) {
+      if (is_stream_set) {
+        // only sync with the stream when we are going to use the data.
+        CUDA_TRY(cudaStreamSynchronize(stream));
+      }
+      JNIEnv* env = cudf::jni::get_jni_env(jvm);
+      handle_buffer(env, current_buffer, current_buffer_written);
+      current_buffer = nullptr;
+      current_buffer_len = 0;
+      current_buffer_data = nullptr;
+      current_buffer_written = 0;
+    }
+  }
+
+  size_t bytes_written() override {
+    return total_written;
+  }
+
+  void set_alloc_size(long size) {
+    this->alloc_size = size;
+  }
+
+private:
+  void rotate_buffer(JNIEnv* env) {
+    if (current_buffer != nullptr) {
+      if (is_stream_set) {
+        // only sync with the stream when we are going to use the data.
+        CUDA_TRY(cudaStreamSynchronize(stream));
```

Comment:
Technically we don't want to synchronize with the entire stream backlog, just the point after we issued the copy.  Really should be synchronizing on an event here that was added to the stream immediately following the copy.  Something to consider in the future.